pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -164,13 +166,73 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
 	if t.pollsToArchive <= 0 || t.persister == nil {
 		return
 	}
-	key := fmt.Sprintf("knownFiles%d", t.archiveIndex)
-	if err := checkpoint.SaveKey(context.Background(), t.persister, metadata.Get(), key); err != nil {
+	if err := t.writeArchive(t.archiveIndex, metadata); err != nil {
 		t.set.Logger.Error("error faced while saving to the archive", zap.Error(err))
 	}
 	t.archiveIndex = (t.archiveIndex + 1) % t.pollsToArchive // increment the index
 }
 
+// readArchive loads data from the archive for a given index and returns a fileset.Filset.
+func (t *fileTracker) readArchive(index int) (*fileset.Fileset[*reader.Metadata], error) {
+	key := fmt.Sprintf("knownFiles%d", index)
+	metadata, err := checkpoint.LoadKey(context.Background(), t.persister, key)
+	if err != nil {
+		return nil, err
+	}
+	f := fileset.New[*reader.Metadata](len(metadata))
+	f.Add(metadata...)
+	return f, nil
+}
+
+// writeArchive saves data to the archive for a given index and returns an error, if encountered.
+func (t *fileTracker) writeArchive(index int, rmds *fileset.Fileset[*reader.Metadata]) error {
+	key := fmt.Sprintf("knownFiles%d", index)
+	return checkpoint.SaveKey(context.Background(), t.persister, rmds.Get(), key)
+}
+
+// FindFiles goes through archive, one fileset at a time and tries to match all fingerprints against that loaded set.
+func (t *fileTracker) FindFiles(fps []*fingerprint.Fingerprint) []fileset.Matchable {
+	// To minimize disk access, we first access the index, then review unmatched files and update the metadata, if found.
+	// We exit if all fingerprints are matched.
+
+	mostRecentIndex := util.Mod(t.archiveIndex-1, t.pollsToArchive)
+	matchedMetadata := make([]fileset.Matchable, len(fps))
+	indices := make(map[int]bool) // Track fp indices of original fps slice
+
+	for i := 0; i < len(fps); i++ {
+		indices[i] = true
+	}
+
+	// continue executing the loop until either all records are matched or all archive sets have been processed.
+	for i := 0; i < t.pollsToArchive && len(indices) > 0; i, mostRecentIndex = i+1, util.Mod(mostRecentIndex-1, t.pollsToArchive) {
+		modified := false
+		data, err := t.readArchive(mostRecentIndex) // we load one fileset atmost once per poll
+		if err != nil {
+			t.set.Logger.Error("error while opening archive", zap.Error(err))
+			continue
+		}
+		for index := range indices {
+			if md := data.Match(fps[index], fileset.StartsWith); md != nil {
+				// update the matched metadata for this index
+				matchedMetadata[index] = md
+				delete(indices, index)
```

Comment:
Using a map and deleting keys from it may be marginally more performant but I don't buy that it's worth the complexity. Just loop over the result slice and only attempt to match an item if the corresponding result is nil.

Reply:
Thanks for pointing this out. I've updated the code.